conanfile.py

```
@@ -49,11 +40,9 @@ def configure(self):
 
     def requirements(self):
         self.requires("librw/{}".format(self.version))
-        self.requires("mpg123/1.26.4")
         if self.options.audio == "openal":
             self.requires("openal/1.21.0")
-        elif self.options.audio == "miles":
-            self.requires("miles-sdk/{}".format(self.version))
```

Please also keep this.

---

Why? Miles doesn't need any OS-level package, nor any it's submodule, why do we need cmake files for it?

---

also if you haven't noticed, all external things that falls under this definition(don't have OS module dependency and not submodule) now under /src/external, not /vendor. So they're now *part* of source.

---

e.g. for mingw. This repo only provides binaries for MSVC.

I have added a conan recipe + cmake build script to https://github.com/withmorten/re3mss to rebuild the miles import library.
This is also useful for purists who want to build everything themselve.

---

> also if you haven't noticed, all external things that falls under this definition(don't have OS module dependency and not submodule) is now under /src/external, not /vendor. So they're now _part_ of source.

You can supply vendored versions of third-party libraries to make it easy for developers.
That does not mean that you just break building with system packages.
If I follow your logic, then we should also remove libsndfile, openal, ogg, ... from conan because it is in the external subfolder.

---

It isn't making it easy, it's making it hard. Having a cmake file for just for one header library doesn't make sense.

Also ogg, opus, opusfile are **still** in /vendor, because they're used by Linux too, they have OS-level dependencies, and they're submodules.

Whereas libsndfile are now in /src/external, because this folder is only for Windows, and nor it needs any OS-level package for Windows nor it's submodule.

---

I know that we need re3mss repo and I don't want to break it. But I want to remove searching for /vendor/milessdk. I will see what can I do.

---

> It isn't making it easy, it's making it hard. Having a cmake file for just for one header library doesn't make sense.

It makes perfect sense in cmake.
It's not "just" one header file. There is also an import library.

---

> I know that we need re3mss repo and I don't want to break it. But I want to remove searching for /vendor/milessdk. I will see what can I do.

What do you mean by "searching for /vendor/milessdk"?
The current `FIndMilesSDK.cmake` looks in `MilesSDK_DIR`.

---

> It makes perfect sense in cmake.
It's not "just" one header file. There is also an import library.

I don't want that, sorry.

> What do you mean by "searching for /vendor/milessdk"?
The current FIndMilesSDK.cmake looks in MilesSDK_DIR.

I mean we don't need `find_library`, `find_path` etc. because it will always be in /src/external in current proposal. Because
1- Real SDK and headers will never be officialy released
2- It will never get updated
3- It will no way be submodule due to there is no official repo
4- It will never require a package installed on OS

Thus, I and I think a few more people on the team see no point on naming this as an module.
I'm aware that on MinGW we need to fetch re3mss repo and build it, and I know that I should keep it.
So I'm requesting your help on how can I accomplish this. Do you think just removing `find_library` and `find_path` will be enough?

---

Ok, so you don't want to "hide" miles as an external dependency in the conan recipe and just let the cmake build system handle it itself. Correct?

For building re3 with mingw, we need access to the sources of withmorton's miles sdk.
So you should add it as a git submodule or just copy its files to some `vendor` subdirectory.

In the cmake build system, you can then include re3mss a subdirectory, and link to the re3mss target.
e.g.
```cmake
if(${PROJECT}_AUDIO STREQUAL "MSS")
    add_subdirectory("${PROJECT_SOURCE_DIR}/external/re3mss" re3mss)
    target_compile_definitions(${EXECUTABLE} PRIVATE AUDIO_MSS)
    target_link_libraries(${EXECUTABLE PRIVATE MilesSDK::MilesSDK)
endif()
```

I will open a small pr at @withmorten 's re3mss repo to make `MilesSDK::MilesSDK` available + to don't always install its files.

---

I've opened https://github.com/withmorten/re3mss/pull/3

---

Thank you, but I remember like we already have a code that fetches that repo and compiles it, am i wrong?

---

No, in the conan github ci workflow, a conan package is created from a checked-out re3mss repo.
The re3 conan recipe then uses that re3mss conan package.

The other ci workflows are MSVC specific and use the import library that is available in this repo.

Adding code to cmake to [fetch a subproject](https://cmake.org/cmake/help/latest/module/FetchContent.html) is possible but would make it not reproducible. Downloading something at build time is not really something nice.

---

I see, thank you. withmorten just merged the PR, would you like the adapt main repo for it?

---

So would you be okay with re3mss as a git module?

---

I think that'd be okay, but I'd add a precompiled .lib for MSVC builds at least before that.

---

Yup. I'll add the submodule as `vendor/milessdk/sources`.

---

And I would appreciate if you can move vendor/milessdk content to src/external/milessdk.

---

Currently, `vendor/milessdk` has the following tree:
```
vendor/milessdk
├── include
│   └── mss.h
└── lib
      └── mss32.lib
```
Note how it contains an import library.

I propose to change it to:
```
vendor/milessdk
├── include
│   └── mss.h
├── lib
│   └── mss32.lib
└── source
    ├── CMakeLists.txt
    ├── re3mss
    │   ├── mss32.c
    │   ├── mss.h
    │   ├── re3mss.vcxproj
    │   └── re3mss.vcxproj.filters
    ├── re3mss.sln
    └── README.md
```

Note also that `vendor/librw` only contains sources, no libraries.
By your logic, `vendor/librw` should be moved to `src/external/librw`.


I think you want the following:
- keep the import library in vendor/milessdk but remove the header
- add re3mss as a submodule in `src/external/milessdk` and use the header from that location

---

Can you check if I need to do more in CMakeLists here? I think "..\include" needs to be added as include dir, but I'm a total cmake noob ...

https://github.com/withmorten/re3mss/tree/test

---

@withmorten 
The cmake script needs a little change:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index b3d7b55..8f7ecd8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -20,7 +20,7 @@ add_library(re3mss SHARED
 )
 add_library(MilesSDK::MilesSDK ALIAS re3mss)
 target_include_directories(re3mss
-    PUBLIC "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/re3mss>"
+    PUBLIC "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>"
     PUBLIC "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>"
 )
 
```

---

My proposal is moving vendor/milessdk to src/external/milessdk, and adding vendor/re3mss as a submodule. Me and withmorten both think header should be located in src/external/milessdk.

I don't see it as sources/libraries seperation, I evaluate where it should belong by checking these;
- does it need any package on OS?
- is it submodule?

since both of them isn't met, I think it belongs to src/external.

---

Btw, if the test branch becomes master and if re3 is supposed to use the import library of re3mss.
Then I think re3mss should be kept as a pure git submodule as `vendor/re3mss`

---

So `rm -rf vendor/milessdk` and everything to `src/external/milessdk`?

---

Yes

---

Okay that fixed it - I have a question though - cmake currently only builds x64 windows anyway right? since there is no miles library for 64bit for miles 6, this could only be used for x86 windows builds anyway - so the cmakelists in re3mss should also generate 32bit configs.

---

Well, cmake can only build for one compiler/arch/os at a time.
This means with a 32-bit compiler, 32-bit executables/libraries will be created.
With a 64-bit compiler, 64-bit executables will be generated.

There is no robust way in cmake to force build a 32-bit executable.
With MSVC, this would require a lot of different compiler flags then mingw.
If the mingw toolchain is build without multilib support, then this is impossible.

What re3mss's cmake script is currently doing is create a `mss64` import library.
The alternative is to fail hard.

---

Hm, so I tried a bit and couldn't get cmake to generate an x86 project, even `cmake -A x86 -G "Visual Studio 16 2019" .` fails ... even after running vcvars32.bat. Well in any case the .lib can be used for potential 32bit cmake MSVC builds anyway (btw we need to heavily modify the project settings there, currently the RelWithDebInfo builds are basically without optimization, or rather, no inling and cache-unfriendly jumptable in the beginning like a Debug build would have), so I suppose it's still useful for mingw32 builds.

---

I think you need to do `cmake -G "Visual Studio 16 2019" -A Win32`.
See https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2016%202019.html#platform-selection

---

Yes, that worked!

---

Check this branch: https://github.com/madebr/re3/tree/milessdk_gitmodule
Its github CI succeeded: https://github.com/madebr/re3/actions?query=branch%3Amilessdk_gitmodule

I used the `test` branch on the re3mss repo.